pkg/controllers/machinesetmigration/machineset_migration_controller.go

```diff
@@ -0,0 +1,356 @@
+		// For requesting pausing of a MAPI resource, it is sufficient to switch the spec.AuthoritativeAPI field on the MAPI resource.
+		// which is already done before this code runs in this controller.
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// For requesting pausing of a CAPI resource, set the paused annotation on it.
+		// The spec.AuthoritativeAPI is set to MachineAPI, meaning that the old authoritativeAPI was ClusterAPI.
+		// So Check if the ClusterAPI resource has the paused annotation, otherwise set it.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: r.CAPINamespace, Name: ms.Name}, capiMachineSet); err != nil {
+			return false, fmt.Errorf("failed to get Cluster API machine set: %w", err)
+		}
+
+		if !annotations.HasPaused(capiMachineSet) {
+			capiMachineSetCopy := capiMachineSet.DeepCopy()
+			annotations.AddAnnotations(capiMachineSet, map[string]string{capiv1beta1.PausedAnnotation: ""})
+			if err := r.Patch(ctx, capiMachineSet, client.MergeFrom(capiMachineSetCopy)); err != nil {
+				return false, fmt.Errorf("failed to patch Cluster API machine set: %w", err)
+			}
+
+			updated = true
+		}
+
+		// InfraMachineTemplate doesn't have a reconciler and thus it doesn't need pausing.
+		// The only provider we are interested in, that reconciles the inframachinetemplate, is ibmcloud/powervs
+		// which only updates its status
+		// see: https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/controllers/ibmpowervsmachinetemplate_controller.go
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	return updated, nil
+}
+
+func (r *MachineSetMigrationReconciler) isSynchronizedGenerationMatchingOldAuthoritativeAPIGeneration(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) (bool, error) {
+	//nolint:wsl,exhaustive
+	switch mapiMachineSet.Spec.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		// ClusterAPI is the desired authoritative API but
+		// MachineAPI is currently still the authoritative one.
+		// Check MachineAPI generation against the synchronized one.
+		return mapiMachineSet.Status.SynchronizedGeneration == mapiMachineSet.Generation, nil
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// MachineAPI is the desired authoritative API but
+		// ClusterAPI is currently still the authoritative one.
+		// Check ClusterAPI generation against the synchronized one.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: r.CAPINamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return false, fmt.Errorf("failed to get Cluster API machine set: %w", err)
+		}
+
+		// Given the CAPI infra machine template is immutable
+		// we do not check for its generation to be synced up with the generation of the MAPI machine set.
+		return (mapiMachineSet.Status.SynchronizedGeneration == capiMachineSet.Generation), nil
+	default:
```

Comment:
Is this possible? It would allow you to remove exhaustive and also help if we ever did add new values, since exhaustive would point out we haven't made a choice for the new value
```suggestion
	case machinev1beta1.Migrating, default:
```

Reply:
Dropped the `default` and added that one.